lib/src/rules/use_key_in_widget_constructors.dart

```diff
@@ -0,0 +1,95 @@
+// Copyright (c) 2019, the Dart project authors. Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+import 'package:analyzer/dart/ast/ast.dart';
+import 'package:analyzer/dart/ast/visitor.dart';
+import 'package:analyzer/dart/element/type.dart';
+
+import '../analyzer.dart';
+import '../util/dart_type_utilities.dart';
+import '../util/flutter_utils.dart';
+
+const _desc = r'Use key in widget constructors.';
+
+const _details = r'''
+**DO** use key in widget constructors.
+
+It's a good pratice to expose the ability to provide a key when creating public
+widgets.
+
+**BAD:**
+```
+class MyPublicWidget extends StatelessWidget {
+}
+```
+
+**GOOD:**
+```
+class MyPublicWidget extends StatelessWidget {
+  MyPublicWidget({Key key}) : super(key: key);
+}
+```
+''';
+
+class UseKeyInWidgetConstructors extends LintRule implements NodeLintRule {
+  UseKeyInWidgetConstructors()
+      : super(
+            name: 'use_key_in_widget_constructors',
+            description: _desc,
+            details: _details,
+            group: Group.errors);
+
+  @override
+  void registerNodeProcessors(NodeLintRegistry registry,
+      [LinterContext context]) {
+    final visitor = _Visitor(this);
+    registry.addClassDeclaration(this, visitor);
+    registry.addConstructorDeclaration(this, visitor);
+  }
+}
+
+class _Visitor extends SimpleAstVisitor<void> {
+  final LintRule rule;
+
+  _Visitor(this.rule);
+
+  @override
+  void visitClassDeclaration(ClassDeclaration node) {
+    var classElement = node.declaredElement;
+    if (classElement.isPublic &&
+        !classElement.isAbstract &&
```

Comment:
Why are abstract classes exempt? And if it's appropriate to allow an abstract class that has no explicit constructor (this condition), should we also allow an abstract class that has an explicit constructor (the condition starting on line 73)? The two conditions are different and I don't understand the reason for the difference. If the difference is appropriate, consider adding a comment explaining it.

Also, there should be some tests covering abstract classes.

Reply:
This condition was added to handle [ObstructingPreferredSizeWidget](https://github.com/flutter/flutter/blob/bc0d35c34e2065fd520f8b0bcd959af86affc20a/packages/flutter/lib/src/cupertino/page_scaffold.dart#L203-L209) that is use only once as an interface in flutter repo. I will take a closer look and see if it's worth to keep this exception.

Reply:
@a14n: did you take a look at this?

Reply:
yes, I change the rule to only look at widgets extending Widget (not implementing)